kb - tips - Long running Callbacks

erikfTW opened the pull request:
This PR includes Long running callbacks tip.

Here we have the original tips: 

- [Long running Callbacks](https://www.taipy.io/tips/long-running-callbacks/)

erikfTW committed:
doc/kb/tips/Long running Callbacks

jrobinAV committed:
Merge branch 'develop' into doc/kb/tips/long_running_callbacks

jrobinAV committed:
Merge branch 'develop' into doc/kb/tips/long_running_callbacks

jrobinAV committed:
moving to dedicated folder

jrobinAV commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+# Long Running Callbacks
+
+Tasks in web applications can take a lot of time, which can lead to problems with communication between 
```
I am concerned about the term "Task" here, which does not refer to the Taipy Task object. Can it be confusing for readers? @erikfTW @FabienLelaquais @FlorianJacta @RymMichaut

This comment applies to all the task usages on this page.

Alternative wording could be function, process, callback, etc.
What do you think?

FabienLelaquais replied:
I completely agree.
We must find synonyms all over the place.
Here that could be replacing "Tasks in web applications" by "Executing a user-define piece of code"

FlorianJacta replied:
Agree! However, this is the terminology of the [doc](https://docs.taipy.io/en/develop/manuals/gui/callbacks/#long-running-callbacks). It should be changed there as well. The goal was to be consistent with the user man.

jrobinAV replied:
A new ticket was created on taipy-doc for user manual:
https://github.com/Avaiga/taipy-doc/issues/603
We can proceed and apply the change on this page. We will take care of the user manual later.

jrobinAV commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+# Long Running Callbacks
```
No need for a H1 title here. The nav bar title will be added automatically, so you will have 2 titles.
```suggestion
```

jrobinAV commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+This article talks about the idea of long-running callbacks in Taipy, gives examples of how to use them, 
+and shows how they make the user experience better.
+
+## Example 1: Executing a Heavy Task in the Background
```
If we remove the H1 title, we can change the title level accordingly.
```suggestion
# Example 1: Executing a Heavy Task in the Background
```

FabienLelaquais replied:
May be not though.
MkDocs adds a H1-like thing, and using a H2 section makes the section title smaller, which is probably what we want.

jrobinAV replied:
Oh... I created dozens of comments in other PRs about that.
🤣🤣🤣🤣
 
We need to make a global decision, I guess.

jrobinAV commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+It sets up a separate background thread to run the `heavy_function()`, allowing the rest of the application to continue running. 
+The `on_action` function gets activated by a user action, like clicking a button, for instance.
+
+## Example 2: Monitoring Task Status
```
```suggestion
# Example 2: Monitoring Task Status
```

FabienLelaquais replied:
"\`heavy_function()\`" => "\*heavy_function()\*"
"The \`on_action\` function" -> "The \`on_action\` callback"

jrobinAV replied:
Wrong line comment, no?

jrobinAV commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+This gives you the opportunity to provide the necessary notifications or make updates to the user interface depending 
+on whether the task was successful or not.
+
+## Handling heavy_function results: Updating the Application State
```
```suggestion
# Handling heavy_function results: Updating the Application State
```

jrobinAV commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+the result of the resource-intensive task. 
+This creates a smooth and seamless user experience.
+
+## Example 3: Tracking Task Progress
```
```suggestion
# Example 3: Tracking Task Progress
```

jrobinAV commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+at the set interval, such as every 5 seconds. 
+This allows your user interface to show live updates, keeping the end user informed about the ongoing task.
+
+## Conclusion
```
```suggestion
# Conclusion
```

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+![Long Running Callbacks](long_running_callbacks.png){width=100%}
+
+To deal with this issue, Taipy has a helpful feature called "long-running callbacks" to address this problem. 
```
"To deal with this issue, Taipy has a helpful feature called "long-running callbacks" to address this problem."
->
"To deal with this issue, Taipy has a helpful feature called "long-running callbacks"."

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+![Long Running Callbacks](long_running_callbacks.png){width=100%}
+
+To deal with this issue, Taipy has a helpful feature called "long-running callbacks" to address this problem. 
+These callbacks allow the server to do demanding tasks in the background while still making sure 
```
tasks -> processing

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+To deal with this issue, Taipy has a helpful feature called "long-running callbacks" to address this problem. 
+These callbacks allow the server to do demanding tasks in the background while still making sure 
+the user interface stays quick to respond. 
```
"quick to respond" -> "responsive"

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+To deal with this issue, Taipy has a helpful feature called "long-running callbacks" to address this problem. 
+These callbacks allow the server to do demanding tasks in the background while still making sure 
+the user interface stays quick to respond. 
+This article talks about the idea of long-running callbacks in Taipy, gives examples of how to use them, 
```
idea -> concept

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+These callbacks allow the server to do demanding tasks in the background while still making sure 
+the user interface stays quick to respond. 
+This article talks about the idea of long-running callbacks in Taipy, gives examples of how to use them, 
+and shows how they make the user experience better.
```
"make the user experience better." -> "can improve the user experience."

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+## Example 1: Executing a Heavy Task in the Background
+
+Imagine a situation where a callback starts a task that requires a lot of resources and time to finish. 
```
task -> duty

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+Imagine a situation where a callback starts a task that requires a lot of resources and time to finish. 
+To make this work, we can use a straightforward approach:
+
+![Executing a Heavy Task in the Background](long_running_callbacks_2.png){width=100%}
```
"a Heavy Task" -> "a time-consuming work"

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+![Executing a Heavy Task in the Background](long_running_callbacks_2.png){width=100%}
+
+In the previous example, the Taipy function called `invoke_long_callback()` manages the resource-intensive task. 
```
\`invoke_long_callback()\` => \`invoke_long_callback()^\`

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+## Example 2: Monitoring Task Status
+
+Moreover, you can send notifications to the user's browser or update visual elements depending on the task's result. 
```
"task's result" -> "status of the ongoing process"

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+![Monitoring Task Status](long_running_callbacks_3.png){width=100%}
+
+In this example, we introduce the `heavy_function_status()` function, which is invoked by the `invoke_long_callback()` method. 
```
\`heavy_function_status()\` => \*heavy_function_status()\*
\`invoke_long_callback()\` => \`invoke_long_callback()^\`

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+![Monitoring Task Status](long_running_callbacks_3.png){width=100%}
+
+In this example, we introduce the `heavy_function_status()` function, which is invoked by the `invoke_long_callback()` method. 
+When the task is finished, this function is called. 
```
task -> callback
Add \<br/\> before the line break.

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+In this example, we introduce the `heavy_function_status()` function, which is invoked by the `invoke_long_callback()` method. 
+When the task is finished, this function is called. 
+This gives you the opportunity to provide the necessary notifications or make updates to the user interface depending 
+on whether the task was successful or not.
```
task -> processing

jrobinAV commented:
Sorry, I deleted the branch by mistake. I reopened it and re-opened the PR as well.

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+## Handling heavy_function results: Updating the Application State
+
+To update the State according to the return of the `heavy_function`, you can modify the `heavy_function_status()` function as follows:
```
"the return of the \`heavy_function\`, you can modify the \`heavy_function_status()\` function as follows:"
->
"the returned value from \*heavy_function()\*, you can modify \*heavy_function_status()\* as follows:"

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+![Updating the Application State](long_running_callbacks_4.png){width=100%}
+
+We added a parameter called result, which represents the return value of the `heavy_function()`. When the `heavy_function()` 
```
result -> \*result\*
"the \`heavy_function()\`" -> "\*heavy_function()\*" (twice!)

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+![Updating the Application State](long_running_callbacks_4.png){width=100%}
+
+We added a parameter called result, which represents the return value of the `heavy_function()`. When the `heavy_function()` 
+completes successfully (status is True), 
```
status -> \*status\*

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+We added a parameter called result, which represents the return value of the `heavy_function()`. When the `heavy_function()` 
+completes successfully (status is True), 
+we update the State with the result by assigning it to a state variable (e.g., state.result = result). 
```
"state.result = result" -> "\`state.result = result\`"

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+we update the State with the result by assigning it to a state variable (e.g., state.result = result). 
+This allows you to access the result in other parts of your application or display it to the user as needed.
+
+Make sure that the `heavy_function()` returns a value. For example:
```
"the \`heavy_function()\`" -> "*heavy_function()*"

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+![Updating the Application State](long_running_callbacks_5.png){width=100%}
+
+When you update the State with the result of the `heavy_function()`, you make sure that the user interface shows 
```
"the \`heavy_function()\`" -> "*heavy_function()*"

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+## Example 3: Tracking Task Progress
+
+Occasionally, it's useful to give regular updates on the progress of a long-running task. Taipy's `invoke_long_callback()` 
```
\`invoke_long_callback()\` => \`invoke_long_callback()^\`

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+![Tracking Task Progress](long_running_callbacks_6.png){width=100%}
+
+In the code above, when you include a `period` parameter, the `heavy_function_status()` function will be regularly activated 
```
\`period\` -> \*period\*
\`heavy_function_status()\` -> \*heavy_function_status()\*

FabienLelaquais commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,73 @@
+
+In the code above, when you include a `period` parameter, the `heavy_function_status()` function will be regularly activated 
+at the set interval, such as every 5 seconds. 
+This allows your user interface to show live updates, keeping the end user informed about the ongoing task.
```
task -> work

FabienLelaquais reviewed (dismissed):
A few minor things, including conventions on typography.

FabienLelaquais commented:
I sincerely hope that the code fragments on the web site (where, well, it wasn't perfect) are kept!
It looks like images replaced code blocks, which makes it impossible to copy/paste it.

erikfTW committed:
Apply suggestions from code review

Co-authored-by: Jean-Robin <jeanrobin.medori@avaiga.com>

erikfTW committed:
Update index.md

erikfTW committed:
Merge branch 'develop' into doc/kb/tips/long_running_callbacks

erikfTW commented:
> I sincerely hope that the code fragments on the web site (where, well, it wasn't perfect) are kept! It looks like images replaced code blocks, which makes it impossible to copy/paste it.

I have fixed that and added to code blocks.

erikfTW committed:
docs/kb/tips/long_callbacks/ code block fixes

jrobinAV committed:
Merge branch 'develop' into doc/kb/tips/long_running_callbacks

jrobinAV committed:
Fix navbar

jrobinAV committed:
Remove unused pictures + formatting

jrobinAV committed:
Formatting

jrobinAV committed:
Formatting

jrobinAV commented:
Please review it again. I applied doc conventions + I tried changing the wording on Task.
Not sure abt the result.

RymMichaut committed:
Update index.md

RymMichaut commented:
I feel like this article lacks of explanatory images

jrobinAV committed:
formatting

jrobinAV committed:
Merge branch 'develop' into doc/kb/tips/long_running_callbacks

erikfTW committed:
Merge branch 'develop' into doc/kb/tips/long_running_callbacks

FlorianJacta commented on docs/knowledge_base/tips/long_callbacks/index.md:
```diff
@@ -0,0 +1,123 @@
+
+![Long Running Callbacks](long_running_callbacks.png){width=100%}
+
+This article discusses the concept of long-running callbacks in Taipy, provides usage examples, 
```
tip
